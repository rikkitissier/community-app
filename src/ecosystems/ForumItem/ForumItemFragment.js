import gql from "graphql-tag";

const ForumItemFragment = gql`
	fragment ForumItemFragment on forums_Forum {
		id
		name
		topicCount
		postCount
		hasUnread
		subforums {
			id
			name
			topicCount
			postCount
			hasUnread
			passwordProtected
			passwordRequired
			isRedirectForum
			redirectHits
			url {
				full
			}
			lastPostAuthor {
				photo
			}
			lastPostDate
		}
	}
`;

export default ForumItemFragment;
