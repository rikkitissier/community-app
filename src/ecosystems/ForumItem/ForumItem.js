import React, { Component } from "react";
import { Text, Image, View, StyleSheet, TouchableHighlight, Alert } from "react-native";
import Swipeable from "react-native-swipeable";
import { graphql, compose, withApollo } from "react-apollo";
import gql from "graphql-tag";
import { connect } from "react-redux";
import { withNavigation } from "react-navigation";
import _ from "underscore";

import ForumItemFragment from "./ForumItemFragment";
import Lang from "../../utils/Lang";
import ContentRow from "../../ecosystems/ContentRow";
import { PlaceholderContainer, PlaceholderElement } from "../../ecosystems/Placeholder";
import ForumIcon from "../../atoms/ForumIcon";
import LastPostInfo from "../../ecosystems/LastPostInfo";
import { withTheme } from "../../themes";
import icons from "../../icons";

const MarkForumRead = gql`
	mutation MarkForumRead($id: ID!) {
		mutateForums {
			markForumRead(id: $id) {
				...ForumItemFragment
			}
		}
	}
	${ForumItemFragment}
`;

class ForumItem extends Component {
	constructor(props) {
		super(props);
		this._swipeable = null;
		this._markReadTimeout = null;
		this.onPress = this.onPress.bind(this);
		this.markForumRead = this.markForumRead.bind(this);
	}

	componentWillUnmount() {
		clearTimeout(this._markReadTimeout);
	}

	/**
	 * Event handler for tappig a forum row
	 *
	 * @param 	object 	section 	The section we're rendering
	 * @return 	Component|null
	 */
	onPress() {
		this.props.navigation.navigate({
			routeName: "TopicList",
			params: {
				id: this.props.data.id,
				title: this.props.data.name,
				subtitle: Lang.pluralize(Lang.get("topics"), Lang.formatNumber(this.props.data.topicCount))
			},
			key: `forum_${this.props.data.id}`
		});
	}

	markForumRead() {
		this._swipeable.recenter();
		this._markReadTimeout = setTimeout(async () => {
			try {
				await this.props.client.mutate({
					mutation: MarkForumRead,
					variables: {
						id: this.props.data.id
					},
					optimisticResponse: {
						mutateForums: {
							__typename: "mutate_Forums",
							markForumRead: {
								...this.props.data,
								subforums: _.isArray(this.props.data.subforums) ? this.props.data.subforums.slice() : null,
								hasUnread: false
							}
						}
					}
				});
			} catch (err) {
				console.log(err);

				Alert.alert(Lang.get("error_marking_read"), Lang.get("error_marking_read_desc"), [{ text: Lang.get("ok") }], { cancelable: false });
			}
		}, 500);
	}

	render() {
		const { styles, componentStyles, styleVars } = this.props;

		if (this.props.loading) {
			return (
				<ContentRow>
					<PlaceholderContainer height={60} style={[styles.mrWide, styles.mlWide, styles.mtTight, styles.mbTight]}>
						<PlaceholderElement circle radius={20} left={0} top={styleVars.spacing.tight} />
						<PlaceholderElement
							width={250}
							height={17}
							top={styleVars.spacing.tight}
							left={20 + styleVars.spacing.standard}
							right={20 + styleVars.spacing.veryWide}
						/>
						<PlaceholderElement width={100} height={13} top={25 + styleVars.spacing.tight} left={20 + styleVars.spacing.standard} />
						<PlaceholderElement circle radius={30} right={0} top={0} />
						<PlaceholderElement width={30} height={12} top={33} right={4} />
					</PlaceholderContainer>
				</ContentRow>
			);
		}

		const rightButtons = [
			<TouchableHighlight style={[styles.flex, styles.flexJustifyCenter, styles.swipeItemWrap]} onPress={this.markForumRead}>
				<View style={[styles.flexColumn, styles.flexAlignCenter, styles.swipeItem]}>
					<Image source={icons.CHECKMARK2} style={styles.swipeItemIcon} resizeMode="contain" />
					<Text style={styles.swipeItemText}>Read</Text>
				</View>
			</TouchableHighlight>
		];

		const lastPostInfo = {
			photo: this.props.data.lastPostAuthor ? this.props.data.lastPostAuthor.photo : null,
			date: this.props.data.lastPostDate
		};

		return (
			<Swipeable rightButtons={rightButtons} onRef={ref => (this._swipeable = ref)}>
				<ContentRow style={componentStyles.forumItem} onPress={this.props.onPress || this.onPress}>
					<View style={componentStyles.iconAndInfo}>
						<ForumIcon style={componentStyles.forumIcon} unread={this.props.data.hasUnread} type={this.props.data.isRedirectForum ? "redirect" : "normal"} />
						<View style={componentStyles.forumInfo}>
							<Text style={[styles.itemTitle, componentStyles.forumTitle]} numberOfLines={1}>
								{this.props.data.name}
							</Text>
							{!this.props.data.isRedirectForum && (
								<Text testId="postCount" style={[styles.lightText, styles.standardText]}>
									{Lang.pluralize(Lang.get("posts"), Lang.formatNumber(this.props.data.topicCount))}
								</Text>
							)}
						</View>
					</View>
					{!Boolean(this.props.data.isRedirectForum) && (
						<LastPostInfo style={componentStyles.lastPost} photo={lastPostInfo.photo} photoSize={30} timestamp={lastPostInfo.date} />
					)}
				</ContentRow>
			</Swipeable>
		);
	}
}

const _componentStyles = styleVars => ({
	forumItem: {
		paddingHorizontal: styleVars.spacing.wide,
		paddingVertical: styleVars.spacing.wide,
		flexDirection: "row",
		justifyContent: "space-between",
		alignContent: "stretch",
		alignItems: "center"
	},
	iconAndInfo: {
		flexDirection: "row",
		flex: 1,
		paddingRight: 20
	},
	forumInfo: {
		marginLeft: 9
	},
	forumTitle: {
		lineHeight: 18
	}
});

export default compose(
	withNavigation,
	withApollo,
	withTheme(_componentStyles)
)(ForumItem);

export { ForumItem as TestForumItem }; // For test runner only
