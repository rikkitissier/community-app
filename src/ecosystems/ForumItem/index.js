import ForumItem from './ForumItem';
import ForumItemFragment from './ForumItemFragment';

export { ForumItem as ForumItem };
export { ForumItemFragment as ForumItemFragment };