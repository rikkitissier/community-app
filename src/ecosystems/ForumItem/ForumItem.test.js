import React from "react";
import gql from "graphql-tag";
//import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import { MockedProvider } from "react-apollo/test-utils";
import ForumItem, { TestForumItem } from "./ForumItem";
import ForumItemFragment from "./ForumItemFragment";

const MarkForumRead = gql`
	mutation MarkForumRead($id: ID!) {
		mutateForums {
			markForumRead(id: $id) {
				...ForumItemFragment
			}
		}
	}
	${ForumItemFragment}
`;

const FORUM_DATA = {
	id: 1,
	name: "Test Forum",
	topicCount: 3,
	postCount: 5,
	hasUnread: true,
	passwordProtected: false,
	passwordRequired: false,
	isRedirectForum: false,
	redirectHits: 0,
	url: {
		full: "http://www.test.com"
	},
	lastPostAuthor: {
		photo: "https://via.placeholder.com/150"
	},
	lastPostDate: 1508115202
};

const MOCKS = [
	{
		request: {
			mutation: MarkForumRead,
			variables: {
				id: 1
			}
		},
		result: {
			data: {
				...FORUM_DATA,
				hasUnread: false
			}
		}
	}
];

describe("ForumItem component", () => {
	ForumItem.prototype.componentDidMount = jest.fn(e => e);
	const navigation = { navigate: jest.fn() };

	it("renders with data", () => {	
		const wrapper = shallow(<TestForumItem data={FORUM_DATA} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("renders a loading state", () => {
		const wrapper = shallow(<TestForumItem loading={true} />);
		expect(wrapper.find('PlaceholderContainer').length).toBeGreaterThanOrEqual(1);
	});

	it("calls onPress (when supplied as a prop)", () => {
		const onPress = jest.fn();
		const wrapper = mount(<TestForumItem data={FORUM_DATA} onPress={onPress} />);

		wrapper.find('ContentRow').first().props().onPress();
		expect(onPress).toHaveBeenCalledTimes(1);
	});

	it("navigates to TopicList in default onPress handler", () => {
		const wrapper = shallow(<TestForumItem data={FORUM_DATA} navigation={navigation} />);
		wrapper.find('ContentRow').first().props().onPress();
		const navParams = navigation.navigate.mock.calls[0][0];

		expect(navParams.routeName).toBe('TopicList');
		expect(navParams.params).toHaveProperty('id', 1);
	});

	it("shows correct post count", () => {
		const wrapper = shallow(<TestForumItem data={FORUM_DATA} />);
		const postCount = wrapper.findWhere(n => n.prop('testId') === 'postCount').first();

		expect(postCount.prop('children').trim()).toBe('posts(8)');
	});

	it("does not show LastPostInfo for redirect forums", () => {
		const isRedirectData = {
			...FORUM_DATA,
			isRedirectForum: true
		};

		const wrapper = shallow(<TestForumItem data={isRedirectData} />);
		expect(wrapper.find('LastPostInfo').length).toBe(0);
	});

	it("has Mark Read swipeable button for normal forums", () => {
		TestForumItem.prototype.markForumRead = jest.fn();
		const wrapper = shallow(<TestForumItem data={FORUM_DATA} />);
		const rightButtons = wrapper.find('Swipeable').first().prop('rightButtons');
		const spy = jest.spyOn(TestForumItem.prototype, 'markForumRead');

		expect(rightButtons).toBeInstanceOf(Array);
		expect(rightButtons).toHaveLength(1);
		rightButtons[0].props.onPress();		
		expect(spy).toHaveBeenCalledTimes(1);
	});
});
