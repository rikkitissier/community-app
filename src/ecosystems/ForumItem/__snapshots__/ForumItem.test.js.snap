// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForumItem component renders with data 1`] = `
<Swipeable
  leftActionActivationDistance={125}
  leftActionReleaseAnimationConfig={null}
  leftActionReleaseAnimationFn={null}
  leftButtonWidth={75}
  leftButtons={null}
  leftButtonsActivationDistance={75}
  leftButtonsCloseReleaseAnimationConfig={null}
  leftButtonsCloseReleaseAnimationFn={null}
  leftButtonsOpenReleaseAnimationConfig={null}
  leftButtonsOpenReleaseAnimationFn={null}
  leftContent={null}
  onLeftActionActivate={[Function]}
  onLeftActionComplete={[Function]}
  onLeftActionDeactivate={[Function]}
  onLeftActionRelease={[Function]}
  onLeftButtonsActivate={[Function]}
  onLeftButtonsCloseComplete={[Function]}
  onLeftButtonsCloseRelease={[Function]}
  onLeftButtonsDeactivate={[Function]}
  onLeftButtonsOpenComplete={[Function]}
  onLeftButtonsOpenRelease={[Function]}
  onPanAnimatedValueRef={[Function]}
  onRef={[Function]}
  onRightActionActivate={[Function]}
  onRightActionComplete={[Function]}
  onRightActionDeactivate={[Function]}
  onRightActionRelease={[Function]}
  onRightButtonsActivate={[Function]}
  onRightButtonsCloseComplete={[Function]}
  onRightButtonsCloseRelease={[Function]}
  onRightButtonsDeactivate={[Function]}
  onRightButtonsOpenComplete={[Function]}
  onRightButtonsOpenRelease={[Function]}
  onSwipeComplete={[Function]}
  onSwipeMove={[Function]}
  onSwipeRelease={[Function]}
  onSwipeStart={[Function]}
  rightActionActivationDistance={125}
  rightActionReleaseAnimationConfig={null}
  rightActionReleaseAnimationFn={null}
  rightButtonWidth={75}
  rightButtons={
    Array [
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
              "justifyContent": "center",
              "paddingLeft": 20,
            },
            Object {
              "backgroundColor": "#009BA2",
            },
          ]
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
            }
          }
        >
          Mark Read
        </Text>
      </TouchableHighlight>,
    ]
  }
  rightButtonsActivationDistance={75}
  rightButtonsCloseReleaseAnimationConfig={null}
  rightButtonsCloseReleaseAnimationFn={null}
  rightButtonsOpenReleaseAnimationConfig={null}
  rightButtonsOpenReleaseAnimationFn={null}
  rightContent={null}
  swipeReleaseAnimationConfig={
    Object {
      "duration": 250,
      "easing": [Function],
      "toValue": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
  swipeReleaseAnimationFn={[Function]}
  swipeStartMinDistance={15}
>
  <ContentRow
    onPress={[Function]}
    style={
      Object {
        "alignContent": "stretch",
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      }
    }
  >
    <Component
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "paddingRight": 20,
        }
      }
    >
      <ForumIcon
        unread={true}
      />
      <Component
        style={
          Object {
            "marginLeft": 9,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#171717",
                "fontFamily": "System",
                "fontSize": 17,
                "fontWeight": "600",
              },
              Object {
                "lineHeight": 18,
              },
            ]
          }
        >
          Test Forum
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#657686",
              },
              Object {
                "fontSize": 14,
              },
            ]
          }
          testId="postCount"
        >
          posts(8)
        </Text>
      </Component>
    </Component>
    <LastPostInfo
      photo="https://via.placeholder.com/150"
      photoSize={30}
      timestamp={1508115202}
    />
  </ContentRow>
</Swipeable>
`;
